### ISSUE TITLE
Make newsboat not suspend while bookmarking item

### ISSUE THREAD

**Issue by BombchuLink** [newsboat/newsboat#1556]
With the considered recent editions to bookmarking, it would be nice to have newsboat not suspend while the bookmarking script is running. When grabbing images with bookmark-cmd, I have them auto-open with an image viewing program in the bookmarking script after they are downloaded. However, until I close the image viewing program, newsboat says ``Saving bookmark...`.

I have tried using ``setsid`` and ``disown`` in the bookmarking script to have it no longer suspend newsboat, but it does not work.

Even if you could view after downloading through other methods, it still doesn't work for streaming video files to mpv directly, and would still suspend newsboat until all child processes of bookmark-cmd are closed.

**Comment by Minoru** [comment]
This is [by design](https://newsboat.org/releases/2.23/docs/newsboat.html#_bookmarking):

> If everything goes OK, the program simply exits. In case something goes wrong, the program writes out an error message to stdout as a single line. This error message is then presented to the user from within Newsboat. At the moment, Newsboat doesn’t care about the exit code of the program; only its output is used to determine success.

I do see how this makes your use-case inconvenient, but you should be able to write the script in such a way that it exits while the child processes are still running:

> it still doesn't work for streaming video files to mpv directly, and would still suspend newsboat until all child processes of bookmark-cmd are closed.

How are you spawning them? Look up "double fork", it's a common way for the parent process to exit while child processes are left running.

**Comment by BombchuLink** [comment]
I've spent the last hour attempting many things, such as ``disown``, ``setsid``, ``at``, ``&``, and shoving things into functions and using ``sleep`` to get the processes to double fork, but I haven't gotten anything to work with newsboat. 

I'm just going to assume I'm fairly dumb, and it's not an inherent issue with newsboat. 

```bash
#!/bin/sh
url="$1"
title="$2"
description="$3"
feed_title="$4"

function first_fork() {
process_link &
}

function process_link() {
#actual parsing of link begins here
}

#double fork to prevent hanging
first_fork &
```

This is the closest(?) I've managed to get to an answer?

**Comment by Minoru** [comment]
> I'm just going to assume I'm fairly dumb

Apologies, I didn't mean to imply that! You're all right. The issue *is* quite subtle, and took me quite a few minutes to figure out. The script executes the double fork correctly, but it is still connected to its grandparent process (Newsboat) via `stdout`, which it inherits from its parent (`bookmark-cmd`). Since Newsboat is waiting for the end of output, and `stdout` is considered closed only when the last of its file descriptors is closed, Newsboat ends up blocked by `first_fork` even though they are no longer connected in the process hierarchy.

The solution is to redirect `stdout` of the `first_fork` elsewhere:

```
first_fork >/dev/null &
```

You don't need to do this for `stdin` and `stderr` because:

1. Newsboat already closes `stdin` to avoid a circular dependency where `bookmark-cmd` is waiting for more input and Newsboat is waiting for output;
2. Newsboat already redirects `stderr` to /dev/null since it's not interested in that output.

Therefore I'm re-casting this issue as a documentation bug. I think we should describe this use-case in the chapter on bookmarking, and highlight this particular pitfall with `stdout`.

**Comment by Minoru** [comment]
Ah, also, a nitpick: the shebang says "/bin/sh" which is a POSIX shell, but the script itself uses BASH syntax for functions. POSIX doesn't have a `function` keyword ;)

**Comment by BombchuLink** [comment]
> Apologies, I didn't mean to imply that!

Of course not. I just thought it might be relevant to mention such considering my experience with shell isn't stellar.


I have found the double fork isn't even necessary to get newsboat to not hang.

```bash
#!/bin/sh
url="$1"
title="$2"
description="$3"
feed_title="$4"

function process_url() {
#actual parsing of link begins here
}

#re-direct stdout, and fork to prevent newsboat from hanging
process_url >/dev/null &
```

I'm not sure how to do this in a POSIX compliant manner, that doesn't include running a completely separate script. I think the documentation should include a POSIX compliant version if possible?

**Comment by Minoru** [comment]
> I have found the double fork isn't even necessary to get newsboat to not hang.

The first fork is performed by the Newsboat when it spawns a shell to run the script, so yeah, the script itself only has to make one additional fork.

> I think the documentation should include a POSIX compliant version if possible?

That'd be preferable, yes.

> I'm not sure how to do this in a POSIX compliant manner

The only bashism that I see in the updated version is the `function` keyword. To be on a safe side, you can run the script with dash interpreter (should be in repos everywhere), and there's also a tool called shellcheck if you want to really make it perfect.

With that said, perfect is the enemy of the good. So if you want to submit a PR but don't want to bother with POSIX, just don't bother with POSIX ;)
